llmfoundry/tokenizers/tiktoken.py

```diff
@@ -0,0 +1,276 @@
+            except KeyError:
+                pass
+
+        return vocab
+
+    def _tokenize(self, text: str):
+        """Returns a tokenized string.
+
+        Note: We have slightly redefined the expected contract between this method and
+        the _convert_token_to_id method. Normally, this method turns a string, into a list of strings,
+        and then the _convert_token_to_id method turns that list of strings into a list of integers.
+        However, not all vocab indices can be decoded into a string, so instead we just return the integers
+        from this function, and have adjusted the _convert_token_to_id method to handle integers as well as strings.
+        The only use of _tokenize that I could find was in this way, so this _should_ be safe.
+        """
+        tokens = [t for t in self.encoding.encode(text, allowed_special='all')]
+
+        return tokens
+
+    def _convert_token_to_id(self, token: Union[int, str]):
+        """Converts a token (str) in an id using the vocab."""
+        if isinstance(token, int):
+            return token
+        else:
+            return self.encoding.encode(token, allowed_special='all')[0]
+
+    def _convert_id_to_token(self, index: int):
+        """Converts an index (integer) in a token (str) using the vocab."""
+        return self.encoding.decode([index])
+
+    def convert_tokens_to_string(self, tokens: List[str]):
+        """Converts a sequence of tokens (string) in a single string."""
+        return ''.join(tokens)
+
+    def convert_ids_to_tokens(
+            self,
+            ids: Union[int, List[int]],
+            skip_special_tokens: bool = False) -> Union[str, List[str]]:
+        """Converts a single index or a sequence of indices in a token or a.
+
+        sequence of tokens, using the vocabulary and added tokens.
+
+        Args:
+            ids (`int` or `List[int]`):
+                The token id (or token ids) to convert to tokens.
+            skip_special_tokens (`bool`, *optional*, defaults to `False`):
+                Whether or not to remove special tokens in the decoding.
+
+        Returns:
+            `str` or `List[str]`: The decoded token(s).
+        """
+        if isinstance(ids, int):
+            if ids in self.added_tokens_decoder:
+                return self.added_tokens_decoder[ids]
+            else:
+                return self._convert_id_to_token(ids)
+
+        tokens = []
+        current_stream = []
+        for index in ids:
+            if skip_special_tokens and index in self.all_special_ids:
+                continue
+
+            if index in self.added_tokens_decoder:
+                tokens.append(self.encoding.decode(current_stream))
+                current_stream = []
+                tokens.append(self.added_tokens_decoder[index])
+            else:
+                current_stream.append(index)
+
+        if len(current_stream) > 0:
+            tokens.append(self.encoding.decode(current_stream))
```

Comment:
Is tokens a list of single tokens or a list of token sequences?

Reply:
"token" here means "string", so its list of strings

Reply:
Got it, can we add a comment above to say that current_stream is used to collect potentially multiple ids for a single token